MPS Synthesis

ACE07-Sev opened the pull request:
Closes #7650 .

ACE07-Sev committed:
- Fixes #7072

mhucka committed:
Merge branch 'main' into main

ACE07-Sev committed:
- Fixed tester issues.
- Fixed `_parser.py` formatting issues.

ACE07-Sev committed:
- Fixed formatting issues.

ACE07-Sev committed:
- Fixed formatting.

ACE07-Sev committed:
Merge branch 'main' into main

ACE07-Sev committed:
- Added `csx`, `c3sqrtx` and `cp` gates.
- Added testers for `cp`, `c3x`, `c4x`, `csx`, and `c3sqrtx` gates.
- Fixed u0 gate.
- Removed `r`, `ryy`, `cu2`, and `iswap` as they are not part of qelib 1.

ACE07-Sev committed:
- Fixes linting issues.

ACE07-Sev committed:
Merge branch 'quantumlib:main' into main

ACE07-Sev committed:
- Brings back deleted gates [r, ryy, iswap] based on maintainer note.
- Fixes gate definition for `id` to address git-blame note.
- Fixes gate definition for `p` to use ZPowGate.
- Adds rccx and rc3x using `cirq.MatrixGate`.
- Adds missing testers for cry and crz.

ACE07-Sev committed:
- Fixed format and lint issues.

ACE07-Sev committed:
Merge branch 'quantumlib:main' into main

ACE07-Sev committed:
- Added maintainer notes.

ACE07-Sev committed:
Merge branch 'quantumlib:main' into main

ACE07-Sev committed:
Merge branch 'quantumlib:main' into main

ACE07-Sev committed:
Merge branch 'quantumlib:main' into mps

ACE07-Sev committed:
- Initial commit for adding MPS synthesis to `cirq.contrib`.

ACE07-Sev committed:
- Added proper single qubit state case.
- Fixed failed CIs.

ACE07-Sev committed:
- Fixed format issues.

ACE07-Sev committed:
Merge branch 'quantumlib:main' into mps

pavoljuhas commented:
I need to check a few things with colleagues; I will have an update early next week.

pavoljuhas commented on cirq-core/cirq/contrib/mps_synthesis/mps_sequential.py:
```diff
@@ -0,0 +1,317 @@
+    return unitary
+
+
+class Sequential:
```
This may be confusing with the `Sequence` typing class.
Can you rename to MPSSequence or MPSConverter?

Also, please add class docstring.

pavoljuhas commented on cirq-core/cirq/contrib/mps_synthesis/mps_sequential.py:
```diff
@@ -0,0 +1,317 @@
+
+class Sequential:
+    def __init__(
+        self, max_fidelity_threshold: float = 0.95, convention: Literal["lsb", "msb"] = "lsb"
```
The `msb` convention does not seem to be tested nor handled in the code.

Can we for now assume the `lsb` convention only and drop the `convention` argument?

pavoljuhas commented on cirq-core/cirq/contrib/mps_synthesis/mps_sequential.py:
```diff
@@ -0,0 +1,317 @@
+
+        return circuit
+
+    def __call__(
```
Can we convert this to a regular `mps_circuit_from_statevector(statevector, max_num_layers=10, max_fidelity_threshold=0.95)` function and make the `Sequential` class module-internal only?

If so, only the mps_circuit_from_statevector should be exposed in the new `__init__.py` here.

pavoljuhas commented on cirq-core/cirq/contrib/mps_synthesis/mps_sequential.py:
```diff
@@ -0,0 +1,317 @@
+            f"Fidelity: {fidelity:.4f}, "
+            f"Number of qubits: {num_qubits}, "
+            f"Number of layers: {max_num_layers}, "
+        )
```
Seems to be a debugging code - remove?

(If necessary, you can add `mps_fidelity(circuit, statevector)` public function.)

ACE07-Sev replied:
Remove any and all `logger` uses, or just this one?

pavoljuhas commented on cirq-core/cirq/contrib/mps_synthesis/mps_sequential_test.py:
```diff
@@ -0,0 +1,333 @@
+
+    fidelity = np.vdot(cirq.final_state_vector(circuit), state)
+
+    assert np.round(np.abs(fidelity), decimals=6) == 1.0
```
As this should be accurate, can we use `np.testing.assert_allclose(state, cirq.final_state_vector(circuit))` instead?

pavoljuhas reviewed (changes requested):
Please see inline comments.  Otherwise LGTM, thank you for contributing this.

PS: I'd like to loop in @tanujkhattar for more comments after we converge on the initial review.

ACE07-Sev committed:
Merge branch 'quantumlib:main' into mps

ACE07-Sev commented on cirq-core/cirq/contrib/mps_synthesis/mps_sequential_test.py:
```diff
@@ -0,0 +1,333 @@
+def test_compile_with_mps_pass(N: int) -> None:
+    # Generate area-law entangled states for the test
+    state = np.random.rand(2**N) + 1j * np.random.rand(2**N)
+    state /= np.linalg.norm(state)
```
I checked, and `random_superposition` can have volume-law states which are hard to approximate with vanilla MPS (I'll add the sweeping I mentioned before soon, not in this PR though). Is it alright that for this test I keep it as is since the one I have there produces area-law states?

ACE07-Sev committed:
- Changed `gram_schmidt` to `_gram_schmidt` to address the function as internal module only.
- Added assertion to `_gram_schmidt` to raise an error should `matrix` not be square to catch edge cases not known at the time of changing this code.
- Changed class name to `MPSSequential`.
- Added class docstring.
- Removed `convention` from class attributes.
- Added indentation to docstring sections where needed.
- Removed `logger` uses.
- Added `mps_circuit_from_statevector` to `__init__.py` for ease of use.
- Used `cirq.testing.random_superposition` in testers where possible, except `test_compile_area_law_states` which requires strictly area-law entangled states as part of the test.
- Rewrote `test_compile_trivial_state_with_mps_pass` to avoid using qasm and avoid inter-module dependency.
- Used `np.testing.assert_allclose` for about exact matching assertions.

ACE07-Sev commented:
Pushed commit that addressed as many comments as I could. 4 comments remain which need your kind insight.

Main comment remaining is converging on optimal/preferred implementation for `_gram_schmidt`.

ACE07-Sev commented:
@pavoljuhas Greetings sir,

Hope you are well. Have you had a chance to review the new additions?

ACE07-Sev committed:
Merge branch 'main' into mps

pavoljuhas committed:
Reference numpy.typing.NDArray as npt.NDArray

Let us stick to the present pattern.
No change in code function.

pavoljuhas committed:
Replace ValueError with assertion in `_gram_schmidt`

Address missing coverage.
`_gram_schmidt` is module-local and should not be used as public API.

pavoljuhas committed:
Factor out condition which is always-true

Per `_gram_schmidt` type annotation the matrix has True iscomplexobj.

pavoljuhas committed:
Clean up docstrings from type annotations and defaults

These are already present in the function signature.

pavoljuhas committed:
Small tweaks for test readability

No change in the effective code.

pavoljuhas committed:
Merge branch 'main' into mps

pavoljuhas commented on cirq-core/cirq/contrib/mps_synthesis/mps_sequential.py:
```diff
@@ -0,0 +1,317 @@
+    """
+    num_rows, num_columns = matrix.shape
+    unitary = np.zeros((num_rows, num_columns), dtype=np.complex128)
+    orthonormal_basis: list[NDArray[np.complex128]] = []
```
I see, let us then keep it as is for the initial implementation.

pavoljuhas reviewed (approved):
I have pushed in a couple of small clean ups which should unblock the failing coverage.  Looks OK for my part, but I'd like to have another check by @tanujkhattar.

Tanuj - could you please give this a quick look?

pavoljuhas committed:
Typo

ACE07-Sev committed:
Merge branch 'main' into mps

ACE07-Sev committed:
Merge branch 'main' into mps

tanujkhattar commented on cirq-core/cirq/contrib/mps_synthesis/mps_sequential.py:
```diff
@@ -0,0 +1,313 @@
+    Notes
+    -----
+    This implementation provides slightly better fidelity compared to the Schmidt
+    decomposition approach used in scipy.linalg.qr for our use case. If you find
```
Can you add some more details on why this is the case?

mhucka commented:
@tanujkhattar Gentle ping …

ACE07-Sev committed:
Merge branch 'main' into mps

ACE07-Sev committed:
Merge branch 'main' into mps

pavoljuhas reviewed (changes requested):
Per feedback from wjhuggins this is not yet working well enough to be included; for details, please see https://github.com/quantumlib/Cirq/pull/7752#issuecomment-4051396959.

Can you update the PR to address these comments?

ACE07-Sev committed:
Merge branch 'quantumlib:main' into mps
